python/pyiceberg/avro/reader.py

```diff
@@ -238,41 +249,50 @@ def skip(self, decoder: BinaryDecoder) -> None:
             return self.option.skip(decoder)
 
 
-class StructProtocolReader(Reader):
-    create_struct: Callable[[], StructProtocol]
-    fields: Tuple[Tuple[Optional[int], Reader], ...]
+class StructReader(Reader):
+    field_readers: Tuple[Tuple[Optional[int], Reader], ...]
+    create_struct: Type[StructProtocol]
+    struct: Optional[StructType]
 
-    def __init__(self, fields: Tuple[Tuple[Optional[int], Reader], ...], create_struct: Callable[[], StructProtocol]):
-        self.create_struct = create_struct
-        self.fields = fields
-
-    def create_or_reuse(self, reuse: Optional[StructProtocol]) -> StructProtocol:
-        if reuse:
-            return reuse
-        else:
-            return self.create_struct()
+    def __init__(
+        self,
+        field_readers: Tuple[Tuple[Optional[int], Reader], ...],
+        create_struct: Optional[Type[StructProtocol]] = None,
+        struct: Optional[StructType] = None,
```

Comment:
Why is this optional? Won't it always be passed?

Reply:
Some tests didn't pass this, I've made it mandatory.